File: android/src/main/java/expo/modules/iosorb/RotatingGlowView.kt
package expo.modules.iosorb

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.BlendMode
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.rotate
import kotlin.math.min

enum class RotationDirection(val multiplier: Double) {
    CLOCKWISE(1.0),
    COUNTER_CLOCKWISE(-1.0)
}

@Composable
fun RotatingGlowView(
    color: Color,
    rotationSpeed: Double = 30.0,
    direction: RotationDirection,
    elapsedTime: Double,
    modifier: Modifier = Modifier,
    blurRadius: Float = 0f
) {
    val safeSpeed = maxOf(0.01, rotationSpeed)
    val rotation = (elapsedTime * safeSpeed * direction.multiplier).toFloat()

    Canvas(modifier = modifier.fillMaxSize()) {
        val sizePx = min(size.width, size.height)
        val center = Offset(size.width / 2f, size.height / 2f)

        rotate(rotation, pivot = center) {
            // Main glow circle
            drawCircle(
                color = color,
                radius = sizePx / 2f,
                center = center,
                alpha = 0.8f
            )

            // Cutout circle offset downward (creates crescent effect)
            drawCircle(
                color = Color.Transparent,
                radius = sizePx * 0.655f,  // 1.31 / 2
                center = Offset(center.x, center.y + sizePx * 0.31f),
                blendMode = BlendMode.Clear
            )
        }
    }
}

/**
 * Simplified glow without complex masking - just a soft gradient circle
 */
@Composable
fun SimpleRotatingGlow(
    color: Color,
    rotationSpeed: Double = 30.0,
    direction: RotationDirection,
    elapsedTime: Double,
    alpha: Float = 1f,
    modifier: Modifier = Modifier
) {
    val safeSpeed = maxOf(0.01, rotationSpeed)
    val rotation = (elapsedTime * safeSpeed * direction.multiplier).toFloat()
    // Clamp alpha to valid range 0-1
    val safeAlpha = alpha.coerceIn(0f, 1f)

    Canvas(modifier = modifier.fillMaxSize()) {
        val sizePx = min(size.width, size.height)
        val center = Offset(size.width / 2f, size.height / 2f)

        rotate(rotation, pivot = center) {
            // Draw offset glow to create rotating effect
            drawCircle(
                color = color.copy(alpha = (safeAlpha * 0.9f).coerceIn(0f, 1f)),
                radius = sizePx * 0.4f,
                center = Offset(center.x, center.y - sizePx * 0.15f)
            )
            drawCircle(
                color = color.copy(alpha = (safeAlpha * 0.6f).coerceIn(0f, 1f)),
                radius = sizePx * 0.3f,
                center = Offset(center.x + sizePx * 0.1f, center.y + sizePx * 0.1f)
            )
        }
    }
}

File: ios/Orb/RotatingGlowView.swift
//
//  BackgroundView.swift
//  Prototype-Orb
//
//  Created by Siddhant Mehta on 2024-11-06.
//

import SwiftUI

enum RotationDirection {
    case clockwise
    case counterClockwise

    var multiplier: Double {
        switch self {
        case .clockwise: return 1
        case .counterClockwise: return -1
        }
    }
}

struct RotatingGlowView: View {
    private let color: Color
    private let rotationSpeed: Double
    private let direction: RotationDirection
    private let elapsedTime: Double

    init(color: Color,
         rotationSpeed: Double = 30,
         direction: RotationDirection,
         elapsedTime: Double = 0)
    {
        self.color = color
        self.rotationSpeed = rotationSpeed
        self.direction = direction
        self.elapsedTime = elapsedTime
    }

    var body: some View {
        GeometryReader { geometry in
            let size = min(geometry.size.width, geometry.size.height)
            let safeSpeed = max(0.01, rotationSpeed)
            let rotation = elapsedTime * safeSpeed * direction.multiplier

            Circle()
                .fill(color)
                .mask {
                    ZStack {
                        Circle()
                            .frame(width: size, height: size)
                            .blur(radius: size * 0.16)
                        Circle()
                            .frame(width: size * 1.31, height: size * 1.31)
                            .offset(y: size * 0.31)
                            .blur(radius: size * 0.16)
                            .blendMode(.destinationOut)
                    }
                }
                .rotationEffect(.degrees(rotation))
        }
    }
}
